### PR TITLE
Add Gradient Cache&Recompute into Neural Search

### PR DESCRIPTION
**Pull request by w5688414**
### PR types

+ New features

### PR changes

+ APIs

### Description

+ Add Gradient Cache&Recompute into Neural Search

### CODE REVIEW

File: applications/neural_search/recall/in_batch_negative/README.md [hunk 1]
```diff
@@ -229,6 +229,9 @@ python -u -m paddle.distributed.launch --gpus "0,1,2,3" \
 * `recall_num`: 对 1 个文本召回的相似文本数量
 * `similar_text_pair_file`: 由相似文本对构成的评估集
 * `corpus_file`: 召回库数据 corpus_file
+* `use_recompute`: 使用Recompute策略，用于节省显存，是一种以时间换空间的技术
+* `use_gradient_cache`: 使用Gradient Cache策略，用于节省显存，是一种以时间换空间的技术
+* `chunk_numbers`: 使用Gradient Cache策略的参数，表示的是同一个批次的样本分几次执行
```
Comment (wawltor) [inline applications/neural_search/recall/in_batch_negative/README.md:234]: 建议recompute和gradient_cache的这种分布式能力建议收口到Trainer里面，后续@ZHUI来加入

File: applications/neural_search/recall/in_batch_negative/batch_negative/model.py [hunk 2]
```diff
@@ -76,3 +76,56 @@ def forward(self,
...
+        title_cls_embedding = self.get_pooled_embedding(title_input_ids,
+                                                        title_token_type_ids,
+                                                        title_position_ids,
+                                                        title_attention_mask)
+
+        cosine_sim = paddle.matmul(query_cls_embedding,
+                                   title_cls_embedding,
+                                   transpose_y=True)
+
+        # substract margin from all positive samples cosine_sim()
```
Comment (wawltor) [inline applications/neural_search/recall/in_batch_negative/batch_negative/model.py]: 这里的注释注意统一，首字母大写

File: applications/neural_search/recall/in_batch_negative/batch_negative/model.py [hunk 3]
```diff
@@ -76,3 +76,56 @@ def forward(self,
...
+        margin_diag = paddle.full(shape=[query_cls_embedding.shape[0]],
+                                  fill_value=self.margin,
+                                  dtype=paddle.get_default_dtype())
```
Comment (wawltor) [inline applications/neural_search/recall/in_batch_negative/batch_negative/model.py]: 这里的逻辑比较奇怪，为什么不拿一个变量来赋值，可以直接拿cosine_sim dtype

File: applications/neural_search/recall/in_batch_negative/train_batch_neg.py [hunk 4]
```diff
@@ -161,6 +173,175 @@ def evaluate(model, corpus_data_loader, query_data_loader, recall_result_file,
...
+                loss.backward()
+                # Store gradients
+                all_grads.append(model_reps.grad)
+
+            for sub_batch, CUDA_state, grad in zip(sub_batchs,
+                                                   all_CUDA_rnd_state,
+                                                   all_grads):
+
+                sub_query_input_ids, sub_query_token_type_ids, sub_title_input_ids, sub_title_token_type_ids = sub_batch
+                paddle.framework.random.set_cuda_rng_state(CUDA_state)
```
Comment (wawltor) [inline applications/neural_search/recall/in_batch_negative/train_batch_neg.py:308]: 这里还原cuda的随机种子，是为了还原dropout？

File: applications/neural_search/recall/in_batch_negative/batch_negative/model.py (hunk 2 above)
Comment (w5688414) [reply applications/neural_search/recall/in_batch_negative/batch_negative/model.py]: 已修改

File: applications/neural_search/recall/in_batch_negative/batch_negative/model.py (hunk 3 above)
Comment (w5688414) [reply applications/neural_search/recall/in_batch_negative/batch_negative/model.py]: 我试了一下，效果是一样的，已经改成cosine_sim dtype

File: applications/neural_search/recall/in_batch_negative/train_batch_neg.py (hunk 4 above)
Comment (w5688414) [reply applications/neural_search/recall/in_batch_negative/train_batch_neg.py:308]: 由于Gradient Cache需要计算两次前向，设置随机种子是为了使dropout等随机种子一致，确保模型的中间状态一样，使得2次前向得到的结果一致。

File: applications/neural_search/recall/in_batch_negative/README.md (hunk 1 above)
Comment (w5688414) [reply applications/neural_search/recall/in_batch_negative/README.md:234]: 已经发给 @ZHUI

**Review by wawltor — APPROVED** [review]
LGTM
